components/camel-univocity-parsers/src/main/java/org/apache/camel/dataformat/univocity/Marshaller.java

```diff
@@ -86,7 +88,9 @@ private void writeRow(Exchange exchange, Object row, W writer) throws NoTypeConv
         Map<?, ?> map = convertToMandatoryType(exchange, Map.class, row);
         if (adaptHeaders) {
             for (Object key : map.keySet()) {
-                headers.add(convertToMandatoryType(exchange, String.class, key));

```

Comment:
@dbaldes, this way is slightly different if we are talking about sync'ing read and write together. anyway, @zregvart&@dbaldes as i mentioned and tried and saw as well, there are failing test cases with such change. i think if the headers are modified on the fly (as mentioned in the issue as well [link](https://issues.apache.org/jira/browse/CAMEL-11609?focusedCommentId=16103900&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16103900)), sync is a price to pay because `this.headers` holds the (think it of like the column name of the field) column names and if the rows having different number of columns are run into, `this.headers` matches the right columns as per the tests like below

https://github.com/apache/camel/blob/master/components/camel-univocity-parsers/src/test/java/org/apache/camel/dataformat/univocity/UniVocityCsvDataFormatMarshalTest.java#L75

That's why i made a simple change of sync'in the resource to make sure the thread-safety is here.